### PR TITLE
Generate inspection report on CI failures

### PR DESCRIPTION
**Pull request by bschimke95**
This PR adds steps to the CI to create and collect the inspection report if the integration tests fail. The report is uploaded as an artifact of the test run.

### CODE REVIEW

File: .github/workflows/integration-informing.yaml [hunk 1]
```diff
@@ -80,9 +80,27 @@ jobs:
...
+      - name: Generate inspection report
+        if: failure()
+        run: |
+          sudo k8s/scripts/inspect.sh
+      - name: Upload inspection report artifact
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: inspection-report
+          path: ./inspection-report-*.tar.gz
```
Comment (neoaggelos) [inline .github/workflows/integration-informing.yaml:106]: This needs to run as part of the test harness, as we need the inspection report from the LXC instance(s) (or multipass instances, or the test machines in general).

My recommendation would be: Add a config option `TEST_COLLECT_INSPECTION_REPORTS_DIR`, where we dump inspection reports from each instance on the tests (if not empty). Each inspection report should have the instance name somehow. Then, we would have a step here that `if: failure()`, would upload the directory at the end of the run.

This way, a failed run will give us output that describes what failed (the test output we already have), then we'll also have inspection reports from all instances.

**Review by neoaggelos — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-informing.yaml (hunk 1 above)
Comment (neoaggelos) [reply .github/workflows/integration-informing.yaml:106]: This currently runs on the github runner instance, but this is not where k8s-snap is running for our tests

File: tests/integration/tests/test_util/harness/local.py [hunk 2]
```diff
@@ -22,27 +23,33 @@ class LocalHarness(Harness):
...
             raise HarnessError("failed to wait for snapd seed") from e
 
+        self.instance = self.hostname
         return Instance(self, self.hostname)
 
+
+    def get_instances(self) -> Set[str]:
+        if self.instance is None:
+            return set()
+        return set(self.instance)
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply tests/integration/tests/test_util/harness/local.py:51]: but i think having a `.get_instances()` is probably not something we need

File: .github/workflows/integration-informing.yaml (hunk 1 above)
Comment (bschimke95) [reply .github/workflows/integration-informing.yaml:106]: will do after I validated that stuff works in `integration.yaml`

File: tests/integration/tests/conftest.py [hunk 3]
```diff
@@ -137,6 +163,7 @@ def instances(
     # remove the session_instance. The harness ensures that everything is cleaned up
     # at the end of the test session.
     for instance in instances:
+        _generate_inspection_reports(h)
```
[the remark this answers is not in the document]
Comment (bschimke95) [reply tests/integration/tests/conftest.py]: yes, thanks for the catch

File: tests/integration/tests/test_util/harness/local.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bschimke95) [reply tests/integration/tests/test_util/harness/local.py]: removed `get_instances`

File: tests/integration/tests/conftest.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bschimke95) [reply tests/integration/tests/conftest.py]: I also collect the output of the `inspect.sh` script and combine all steps into `_generate_inspection_report(h, instance_id)` now. I call this twice (after test cleanup and after session cleanup) so it makes sense to have this in a separate function

File: tests/integration/tests/conftest.py [hunk 4]
```diff
@@ -25,6 +25,28 @@ def _harness_clean(h: harness.Harness):
...
+    inspection_path = Path(config.INSPECTION_REPORTS_DIR)
+    result = h.exec(
+        instance_id,
+        ["/snap/k8s/current/k8s/scripts/inspect.sh", "/inspection-report.tar.gz"],
+        capture_output=True,
+        text=True,
+    )
+
+    (inspection_path / instance_id).mkdir(parents=True, exist_ok=True)
+    with open(inspection_path / instance_id / "inspection_report_logs.txt", "w") as f:
```
Comment (neoaggelos) [inline tests/integration/tests/conftest.py:40]: you can also use (path).write_text()

Comment (bschimke95) [reply tests/integration/tests/conftest.py:40]: that's actually pretty cool - thanks TIL!

File: tests/integration/tests/conftest.py [hunk 5]
```diff
@@ -37,8 +37,7 @@ def _generate_inspection_report(h: harness.Harness, instance_id: str):
     )
 
     (inspection_path / instance_id).mkdir(parents=True, exist_ok=True)
-    with open(inspection_path / instance_id / "inspection_report_logs.txt", "w") as f:
-        f.write(result.stdout)
+    Path(inspection_path / instance_id / "inspection_report_logs.txt", "w").write_text(result.stdout)
```
Comment (neoaggelos) [inline tests/integration/tests/conftest.py:40]: you probably do not want the `"w"` there, also no need to add `Path()`, as `<path> / "string"` is a Path object already
```suggestion
    (inspection_path / instance_id / "inspection_report_logs.txt").write_text(result.stdout)
```

File: tests/integration/tests/conftest.py [hunk 6]
```diff
@@ -25,6 +25,29 @@ def _harness_clean(h: harness.Harness):
         h.cleanup()
 
 
+def _generate_inspection_report(h: harness.Harness, instance_id: str):
+    LOG.debug("Generating inspection report for %s", instance_id)
+
+    inspection_path = Path(config.INSPECTION_REPORTS_DIR)
+    result = h.exec(
```
Comment (neoaggelos) [inline tests/integration/tests/conftest.py:32]: maybe ignore failures while retrieving the inspection reports.

eg for the etcd instances, where k8s-snap is not installed. you could add maybe check=False for that

File: tests/integration/tests/conftest.py [hunk 7]
```diff
@@ -25,6 +25,30 @@ def _harness_clean(h: harness.Harness):
...
+    (inspection_path / instance_id).mkdir(parents=True, exist_ok=True)
+    (inspection_path / instance_id / "inspection_report_logs.txt").write_text(
+        result.stdout
+    )
+
+    h.pull_file(
+        instance_id,
+        "/inspection-report.tar.gz",
+        (inspection_path / instance_id / "inspection_report.tar.gz").as_posix(),
+    )
```
Comment (neoaggelos) [inline tests/integration/tests/conftest.py:49]: could that, too, fail if the inspection report was not created successfully?

**Review by neoaggelos — APPROVED** [review]

**Pull request merged**